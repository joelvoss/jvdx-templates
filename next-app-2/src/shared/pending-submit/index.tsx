'use client';

import { useFormStatus } from 'react-dom';
import { Button } from 'react-aria-components';
import { mergeProps } from '@/lib/merge-props';
import styles from './index.module.css';

import type { ButtonProps } from 'react-aria-components';

interface PendingSubmitProps extends ButtonProps {}

export function PendingSubmit(props: PendingSubmitProps) {
	const { pending } = useFormStatus();
	const { children, ...rest } = props;

	const buttonProps = mergeProps(
		rest,
		{ type: 'submit' },
		pending ? { isDisabled: true, className: styles.pending } : null,
	);

	return (
		<Button {...buttonProps}>
			{pending ? (
				<>
					<svg className={styles.svg}>
						<use xlinkHref="#loading" />
					</svg>
					{children}
				</>
			) : (
				children
			)}
		</Button>
	);
}
