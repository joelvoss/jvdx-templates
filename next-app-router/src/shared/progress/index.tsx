/**
 * Global progress instance. Initialized lazily.
 */
let globalProgress: ProgressInstance | null = null;
export function getGlobalProgress() {
	if (!globalProgress) {
		globalProgress = new ProgressInstance({ interval: 200 });
	}
	return globalProgress;
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Global progress bar methods that can be used to update the global progress
 * bar from anywhere in the app.
 */
export const Progress = {
	start: () => getGlobalProgress().start(),
	done: () => getGlobalProgress().done(),
};

////////////////////////////////////////////////////////////////////////////////

/**
 * A progress instance handles the state of a progress bar, e.g. automatically
 * incrementing it when a task is started and stopping it when the task is done.
 */
class ProgressInstance {
	private tickleId: NodeJS.Timeout | null = null;
	private subscriptions: Set<() => void> = new Set();
	private interval: number;
	value: number | null = null;

	constructor(options?: { interval?: number }) {
		this.interval = options?.interval ?? 200;
	}

	subscribe(fn: () => void) {
		this.subscriptions.add(fn);
		return () => this.subscriptions.delete(fn);
	}

	set(value: number | null) {
		this.value = value != null ? Math.min(Math.max(value, 0), 1) : null;
		for (let fn of this.subscriptions) {
			fn();
		}
	}

	start() {
		this.set(0);
		this.trickle();
	}

	done() {
		if (this.tickleId) clearTimeout(this.tickleId);
		this.set(1);
		// NOTE(joel): Reset to `null` after `this.speed` ms.
		this.tickleId = setTimeout(() => {
			this.set(null);
		}, this.interval);
	}

	private trickle() {
		this.tickleId = setTimeout(() => {
			if (this.value == null || this.value > 1) return;

			let step = 0;
			if (this.value >= 0 && this.value < 0.2) {
				step = 0.1;
			} else if (this.value >= 0.2 && this.value < 0.5) {
				step = 0.04;
			} else if (this.value >= 0.5 && this.value < 0.8) {
				step = 0.02;
			} else if (this.value >= 0.8 && this.value < 0.99) {
				step = 0.005;
			}
			// NOTE(joel): Don't go over 99,4%
			this.set(Math.min(this.value + step, 0.994));
			this.trickle();
		}, this.interval);
	}
}
