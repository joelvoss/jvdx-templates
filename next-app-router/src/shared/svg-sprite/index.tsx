export function SvgSprite() {
	return (
		<svg
			className='ðŸ‘»'
			xmlns='http://www.w3.org/2000/svg'
			xmlnsXlink='http://www.w3.org/1999/xlink'
		>
			<symbol id='loading' viewBox='0 0 24 24'>
				<path
					fill='currentColor'
					d='M12,1A11,11,0,1,0,23,12,11,11,0,0,0,12,1Zm0,19a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z'
					opacity='.25'
				/>
				<path
					fill='currentColor'
					d='M10.14,1.16a11,11,0,0,0-9,8.92A1.59,1.59,0,0,0,2.46,12,1.52,1.52,0,0,0,4.11,10.7a8,8,0,0,1,6.66-6.61A1.42,1.42,0,0,0,12,2.69h0A1.57,1.57,0,0,0,10.14,1.16Z'
				/>
			</symbol>

			<symbol id='close' viewBox='0 0 24 24'>
				<path
					fill='currentColor'
					d='M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z'
				/>
			</symbol>

			<symbol id='translate' viewBox='0 0 24 24'>
				<path
					fill='currentColor'
					d='m12.87 15.07-2.54-2.51.03-.03A17.52 17.52 0 0 0 14.07 6H17V4h-7V2H8v2H1v1.99h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11.76-2.04zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2l-4.5-12zm-2.62 7 1.62-4.33L19.12 17h-3.24z'
				/>
			</symbol>

			<symbol id='notification' viewBox='0 0 24 24'>
				<path
					fill='currentColor'
					d='M12 22c1.1 0 2-.9 2-2h-4a2 2 0 0 0 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z'
				/>
			</symbol>

			<symbol id='delete' viewBox='0 0 24 24'>
				<path
					fill='currentColor'
					d='M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z'
				/>
			</symbol>

			<symbol id='edit' viewBox='0 0 24 24'>
				<path
					fill='currentColor'
					d='M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z'
				/>
			</symbol>

			{/* Flags; Source https://flagicons.lipis.dev */}
			<symbol id='flag-german' viewBox='0 0 640 480'>
				<path fill='#ffce00' d='M0 320h640v160H0z' />
				<path fill='#000001' d='M0 0h640v160H0z' />
				<path fill='red' d='M0 160h640v160H0z' />
			</symbol>
			<symbol id='flag-english' viewBox='0 0 640 480'>
				<path fill='#bd3d44' d='M0 0h640v480H0' />
				<path
					stroke='#fff'
					strokeWidth='37'
					d='M0 55.3h640M0 129h640M0 203h640M0 277h640M0 351h640M0 425h640'
				/>
				<path fill='#192f5d' d='M0 0h364.8v258.5H0' />
				<marker id='us-a' markerHeight='30' markerWidth='30'>
					<path fill='#fff' d='m14 0 9 27L0 10h28L5 27z' />
				</marker>
				<path
					fill='none'
					markerMid='url(#us-a)'
					d='m0 0 16 11h61 61 61 61 60L47 37h61 61 60 61L16 63h61 61 61 61 60L47 89h61 61 60 61L16 115h61 61 61 61 60L47 141h61 61 60 61L16 166h61 61 61 61 60L47 192h61 61 60 61L16 218h61 61 61 61 60z'
				/>
			</symbol>
		</svg>
	);
}
