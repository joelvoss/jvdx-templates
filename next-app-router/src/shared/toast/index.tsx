'use client';

import { Toaster } from './toaster';
import { Toast } from './toast';
import {
	ToastQueue as StatelyToastQueue,
	useToastQueue,
} from '@react-stately/toast';

import type { AriaToastRegionProps } from '@react-aria/toast';
import type { ToastOptions } from '@react-stately/toast';
import type { ToastValue } from './toast';

////////////////////////////////////////////////////////////////////////////////

// NOTE(joel): We use a single global toast queue instance for the whole app,
// initialized lazily.
let globalToastQueue: StatelyToastQueue<ToastValue> | null = null;
function getGlobalToastQueue() {
	if (globalToastQueue == null) {
		globalToastQueue = new StatelyToastQueue({
			maxVisibleToasts: 5,
			hasExitAnimation: true,
		});
	}
	return globalToastQueue;
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Global `ToastQueue` object that can be used to add toasts to the
 * global queue outside of React.
 */
export const ToastQueue = {
	info(children: string, options: AddToastOptions = {}) {
		return addToast(children, 'info', options);
	},
	positive(children: string, options: AddToastOptions = {}) {
		return addToast(children, 'positive', options);
	},
	negative(children: string, options: AddToastOptions = {}) {
		return addToast(children, 'negative', options);
	},
};

////////////////////////////////////////////////////////////////////////////////

interface ToastContainerProps extends AriaToastRegionProps {}

/**
 * A ToastContainer renders the queued toasts in an application.
 * It should be placed at the root of the app.
 */
export function ToastContainer(props: ToastContainerProps) {
	let state = useToastQueue(getGlobalToastQueue());
	// if (ref === activeToastContainer && state.visibleToasts.length > 0) {
	if (state.visibleToasts.length > 0) {
		return (
			<Toaster state={state} {...props}>
				{state.visibleToasts.map(toast => (
					<Toast key={toast.key} toast={toast} state={state} />
				))}
			</Toaster>
		);
	}

	return null;
}

////////////////////////////////////////////////////////////////////////////////

interface AddToastOptions extends Omit<ToastOptions, 'priority'> {
	/** A label for the action button within the toast. */
	actionLabel?: string;
	/** Handler that is called when the action button is pressed. */
	onAction?: () => void;
	/** Whether the toast should automatically close when an action is performed. */
	shouldCloseOnAction?: boolean;
}

/**
 * Adds a toast to the global `ToastQueue`.
 */
function addToast(
	children: string,
	variant: ToastValue['variant'],
	options: AddToastOptions = {},
) {
	const queue = getGlobalToastQueue();

	const key = queue.add(
		{
			children,
			variant,
			actionLabel: options.actionLabel,
			onAction: options.onAction,
			shouldCloseOnAction: options.shouldCloseOnAction,
		},
		{
			priority: getPriority(variant, options),
			timeout: getTimeout(options),
			onClose: options.onClose,
		},
	);

	return () => queue.close(key);
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Calculates the timeout for a toast based on the options.
 * Toasts must be visible at least 5s.
 * In addition, actionable toasts cannot be auto dismissed
 * @see https://www.w3.org/WAI/WCAG21/Understanding/timing-adjustable.html
 */
function getTimeout(options: AddToastOptions) {
	return options.timeout && !options.onAction
		? Math.max(options.timeout, 5000)
		: undefined;
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Get the priority of a toast based on its variant and options.
 * If a toast has an action, it should be more important than positive toasts
 * without an action but less important than negative toasts.
 */
function getPriority(variant: ToastValue['variant'], options: AddToastOptions) {
	let priority = 1;
	switch (variant) {
		case 'negative': {
			priority = 5;
			break;
		}
		case 'positive': {
			priority = 2;
			break;
		}
		case 'info': {
			priority = 1;
			break;
		}
	}
	if (options.onAction != null) priority += 2;
	return priority;
}
