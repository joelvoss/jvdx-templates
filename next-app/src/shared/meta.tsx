import Head from 'next/head';

type MetaProps = {
	title?: string;
	description?: string;
	keywords?: string;
	image?: string;
	url?: string;
	author?: string;
};

const fallback: MetaProps = {
	title: 'template-next-app',
	description: 'Next.js Template Application',
	keywords: '@jvdx/core, Next.js, Template',
	image: '/android-chrome-512x512.png',
	url: 'https://github.com/joelvoss/jvdx-core',
	author: 'Joel Voss',
};

export function Meta(props: MetaProps) {
	const content = { ...fallback, ...props };

	return (
		<Head>
			<meta charSet="utf-8" />
			<meta name="viewport" content="width=device-width" />

			{/* Favicon */}
			<link
				rel="apple-touch-icon"
				sizes="180x180"
				href="/apple-touch-icon.png"
			/>
			<link
				rel="icon"
				type="image/png"
				sizes="32x32"
				href="/favicon-32x32.png"
			/>
			<link
				rel="icon"
				type="image/png"
				sizes="16x16"
				href="/favicon-16x16.png"
			/>
			<link rel="manifest" href="/site.webmanifest" />
			<meta name="theme-color" content="#555" />

			{/* Twitter */}
			<meta name="twitter:title" content={content.title} />
			<meta name="twitter:description" content={content.description} />
			<meta name="twitter:image" content={content.image} />

			{/* Facebook */}
			<meta property="og:title" content={content.title} />
			<meta property="og:description" content={content.description} />
			<meta property="og:image" content={content.image} />
			<meta property="og:url" content={content.url} />

			<title>{content.title}</title>
			<meta name="description" content={content.description} />
			<meta name="keywords" content={content.keywords} />
			<meta name="author" content={content.author} />
		</Head>
	);
}
