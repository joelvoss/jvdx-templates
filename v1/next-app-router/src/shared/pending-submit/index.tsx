'use client';

import { useFormStatus } from 'react-dom';
import { mergeProps } from '@/lib/merge-props';
import styles from './index.module.css';

import type { ComponentProps } from 'react';

interface PendingSubmitProps extends ComponentProps<'button'> {}

export function PendingSubmit(props: PendingSubmitProps) {
	const { pending } = useFormStatus();
	const { children, ...rest } = props;

	const buttonProps = mergeProps(
		rest,
		{ type: 'submit' },
		pending ? { disabled: true, className: styles.pending } : null,
	);

	return (
		// NOTE(joel): We cant use React Arias <Button> here, since there is a
		// bug in the current `useFormStatus` implementation that prevents us from
		// using it with. See https://github.com/facebook/react/issues/30368
		<button {...buttonProps}>
			{pending ? (
				<>
					<svg className={styles.svg}>
						<use xlinkHref="#loading" />
					</svg>
					{children}
				</>
			) : (
				children
			)}
		</button>
	);
}
