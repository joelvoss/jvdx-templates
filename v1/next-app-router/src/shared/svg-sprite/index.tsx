export function SvgSprite() {
	return (
		<svg
			className="ðŸ‘»"
			xmlns="http://www.w3.org/2000/svg"
			xmlnsXlink="http://www.w3.org/1999/xlink"
		>
			<symbol id="loading" viewBox="0 0 24 24">
				<path
					fill="currentColor"
					d="M12,1A11,11,0,1,0,23,12,11,11,0,0,0,12,1Zm0,19a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
					opacity=".25"
				/>
				<path
					fill="currentColor"
					d="M10.14,1.16a11,11,0,0,0-9,8.92A1.59,1.59,0,0,0,2.46,12,1.52,1.52,0,0,0,4.11,10.7a8,8,0,0,1,6.66-6.61A1.42,1.42,0,0,0,12,2.69h0A1.57,1.57,0,0,0,10.14,1.16Z"
				/>
			</symbol>

			<symbol id="close" viewBox="0 0 24 24">
				<path
					fill="currentColor"
					d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
				/>
			</symbol>

			<symbol id="translate" viewBox="0 0 24 24">
				<path
					fill="currentColor"
					d="m12.87 15.07-2.54-2.51.03-.03A17.52 17.52 0 0 0 14.07 6H17V4h-7V2H8v2H1v1.99h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11.76-2.04zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2l-4.5-12zm-2.62 7 1.62-4.33L19.12 17h-3.24z"
				/>
			</symbol>

			<symbol id="notification" viewBox="0 0 24 24">
				<path
					fill="currentColor"
					d="M12 22c1.1 0 2-.9 2-2h-4a2 2 0 0 0 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
				/>
			</symbol>

			<symbol id="check-circle" viewBox="0 0 24 24">
				<path
					fill="currentColor"
					d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
				/>
			</symbol>

			<symbol id="error" viewBox="0 0 24 24">
				<path
					fill="currentColor"
					d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
				/>
			</symbol>

			{/* Flags; Source https://flagicons.lipis.dev */}
			<symbol id="flag-german" viewBox="0 0 640 480">
				<path fill="#ffce00" d="M0 320h640v160H0z" />
				<path fill="#000001" d="M0 0h640v160H0z" />
				<path fill="red" d="M0 160h640v160H0z" />
			</symbol>
			<symbol id="flag-english" viewBox="0 0 640 480">
				<path fill="#bd3d44" d="M0 0h640v480H0" />
				<path
					stroke="#fff"
					strokeWidth="37"
					d="M0 55.3h640M0 129h640M0 203h640M0 277h640M0 351h640M0 425h640"
				/>
				<path fill="#192f5d" d="M0 0h364.8v258.5H0" />
				<marker id="us-a" markerHeight="30" markerWidth="30">
					<path fill="#fff" d="m14 0 9 27L0 10h28L5 27z" />
				</marker>
				<path
					fill="none"
					markerMid="url(#us-a)"
					d="m0 0 16 11h61 61 61 61 60L47 37h61 61 60 61L16 63h61 61 61 61 60L47 89h61 61 60 61L16 115h61 61 61 61 60L47 141h61 61 60 61L16 166h61 61 61 61 60L47 192h61 61 60 61L16 218h61 61 61 61 60z"
				/>
			</symbol>
		</svg>
	);
}
