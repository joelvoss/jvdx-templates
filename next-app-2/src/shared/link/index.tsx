'use client';

import { useMemo } from 'react';
import NextLink from 'next/link';
import { usePathname } from 'next/navigation';
import { mergeProps } from '@/lib/merge-props';
import { getGlobalProgress } from '@/shared/progress';
import { formatUrl } from './format-url';
import { addBasePath } from './add-base-path';

import type { LinkProps as NextLinkProps } from 'next/link';
import type { MouseEvent } from 'react';
import type { UrlObject } from 'url';

////////////////////////////////////////////////////////////////////////////////

const ABSOLUTE_URL_REGEX = /^[a-zA-Z][a-zA-Z\d+\-.]*?:/;

////////////////////////////////////////////////////////////////////////////////

export interface LinkProps
	extends React.PropsWithChildren<
		Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, keyof NextLinkProps> &
			NextLinkProps
	> {}

export function Link(props: LinkProps) {
	const progress = getGlobalProgress();
	const pathname = usePathname();

	const { href: hrefProp, as: asProp } = props;

	// NOTE(joel): Since `<Link>` accepts `href` as `string` or `UrlObject`
	//  we have to parse it to be able to decide, if we have to start our
	// progress bar. Code ist modified from `next/link` source code.
	// @see https://github.com/vercel/next.js/blob/canary/packages/next/src/client/link.tsx#L454
	let { href, as } = useMemo(() => {
		const resolvedHref = formatStringOrUrl(hrefProp);
		return {
			href: resolvedHref,
			as: asProp ? formatStringOrUrl(asProp) : resolvedHref,
		};
	}, [hrefProp, asProp]);

	if (ABSOLUTE_URL_REGEX.test(as)) {
		href = as;
	} else {
		href = addBasePath(as);
	}

	const linkProps = mergeProps(props, {
		onClick: (event: MouseEvent<HTMLAnchorElement>) => {
			if (shouldTriggerStartEvent(href, event)) progress.start();
			if (props.onClick) props.onClick(event);
		},
		prefetch: false,
		locale: false,
		'data-active': href === pathname,
	});

	return <NextLink {...linkProps} />;
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Formats a URL object or string to a string.
 */
function formatStringOrUrl(urlObjOrString: UrlObject | string) {
	if (typeof urlObjOrString === 'string') {
		return urlObjOrString;
	}
	return formatUrl(urlObjOrString);
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Returns `true` if the link should trigger a start event.
 */
function shouldTriggerStartEvent(href: string, clickEvent?: React.MouseEvent) {
	const current = window.location;
	const target = new URL(href, location.href);
	// NOTE(joel): Modified events -> Fallback to browser behaviour
	if (clickEvent && isModifiedEvent(clickEvent)) return false;
	// NOTE(joel): External URL
	if (current.origin !== target.origin) return false;
	// NOTE(joel): Same URL
	if (current.pathname === target.pathname && current.search === target.search)
		return false;
	return true;
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Returns `true` if the event is a modified click event.
 * Source: https://github.com/vercel/next.js/blob/canary/packages/next/src/client/link.tsx#L180
 */
function isModifiedEvent(event: React.MouseEvent) {
	const eventTarget = event.currentTarget as HTMLAnchorElement | SVGAElement;
	const target = eventTarget.getAttribute('target');
	return (
		(target && target !== '_self') ||
		event.metaKey ||
		event.ctrlKey ||
		event.shiftKey ||
		event.altKey || // triggers resource download
		(event.nativeEvent && event.nativeEvent.which === 2)
	);
}
