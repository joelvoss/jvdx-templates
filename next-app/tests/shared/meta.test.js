/**
 * @jest-environment jsdom
 */

import { render } from '@testing-library/react';
import { Meta } from '@/shared/meta';
import { HeadProvider } from '../jest.setup';

describe('<Meta>', () => {
	test('renders proper HTML', () => {
		// NOTE(joel): Since we're rendering inside <head>, we have to use a special
		// <HeaderProvider> component to flush meta tags in tests.
		const { baseElement } = render(<Meta />, { wrapper: HeadProvider });

		// NOTE(joel): We have to get the document.head ourself based off of the
		// rendered `baseElement`.
		const head = baseElement.parentElement?.firstChild;

		expect(head).toMatchSnapshot();
	});
});
