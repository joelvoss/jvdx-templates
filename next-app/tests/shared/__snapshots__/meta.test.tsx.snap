// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Meta> renders proper HTML 1`] = `
<head>
  <meta
    charset="utf-8"
  />
  <meta
    content="width=device-width"
    name="viewport"
  />
  <link
    href="/apple-touch-icon.png"
    rel="apple-touch-icon"
    sizes="180x180"
  />
  <link
    href="/favicon-32x32.png"
    rel="icon"
    sizes="32x32"
    type="image/png"
  />
  <link
    href="/favicon-16x16.png"
    rel="icon"
    sizes="16x16"
    type="image/png"
  />
  <link
    href="/site.webmanifest"
    rel="manifest"
  />
  <meta
    content="#555"
    name="theme-color"
  />
  <meta
    content="template-next-app"
    name="twitter:title"
  />
  <meta
    content="Next.js Template Application"
    name="twitter:description"
  />
  <meta
    content="/android-chrome-512x512.png"
    name="twitter:image"
  />
  <meta
    content="template-next-app"
    property="og:title"
  />
  <meta
    content="Next.js Template Application"
    property="og:description"
  />
  <meta
    content="/android-chrome-512x512.png"
    property="og:image"
  />
  <meta
    content="https://github.com/joelvoss/jvdx-core"
    property="og:url"
  />
  <title>
    template-next-app
  </title>
  <meta
    content="Next.js Template Application"
    name="description"
  />
  <meta
    content="@jvdx/core, Next.js, Template"
    name="keywords"
  />
  <meta
    content="Joel Voss"
    name="author"
  />
</head>
`;
